### PR TITLE
Support 'raw identifier' syntax

### PR DESCRIPTION
**Pull request by stedolan**
This PR implements part of [RFC 27](https://github.com/ocaml/RFCs/pull/27) by adding support for 'raw identifiers' `\#foo`. The syntax `\#foo` means the same thing as `foo`, but is always an identifier even when `foo` is a keyword. This allows future versions of the language to introduce new keywords in a backwards-compatible way (see the RFC for full details).

The important part of this patch is the small patch to `lexer.mll` which adds the new concrete syntax. The entire rest of the patch is changes to various pretty-printers, to ensure that keywords are printed in appropriately escaped form should they ever occur as variable names, in types, in error messages, and the like.

### CODE REVIEW

File: tools/.depend [hunk 1]
```diff
@@ -160,6 +160,8 @@ ocamlmklib.cmo : \
 ocamlmklib.cmx : \
     ../utils/misc.cmx \
     ../utils/config.cmx
+ocamlmklibconfig.cmo :
+ocamlmklibconfig.cmx :
```
Comment (dra27) [inline tools/.depend]: Small-to-tiny nit - your worktree has an old `tools/ocamlmklibconfig.ml` file knocking around

Comment (stedolan) [reply tools/.depend]: Thanks! I was wondering what was causing alldepend to differ on that machine.

Comment (dra27) [reply tools/.depend]: It's a nuisance that it picks up all files - I often up with `foo.cmx` (or ruder basenames, depending on how frustrating the bug is...) getting picked up, but I don't expect it's worth making `alldepend` more complicated than it already is!

**Comment by stedolan** [comment]
@gasche You expressed interest in this patch at some point, would you be willing / have time to review?

**Comment by gasche** [comment]
Yes, during ICFP last September I realized that I had forgotten about the PR (I liked the RFC and ended up in the awkward position of encouraging you to send a PR about it), and then I quickly forgot about it again.

> You expressed interest in this patch at some point

Yes!

> would [...] have time to review?

No!

But tomorrow we have a meeting about encouraging people to do right by reviewing more stuff, let us try to designate some volunteers there. So many people eager to help, it would be a shame if I stole this opportunity from them.

**Comment by dra27** [comment]
I regret that this was allowed to slip off the radar for 5.0 - it should certainly be in 5.1 and if you're happy to rebase I'm happy to either review it or ensure it gets reviewed in time!

**Comment by stedolan** [comment]
Now rebased!

File: parsing/lexer.mll [hunk 2]
```diff
@@ -398,12 +401,16 @@ rule token = parse
...
+  | "?" raw_ident_escape (lowercase identchar * as name) ':'
+      { OPTLABEL name }
   | "?" (lowercase identchar * as name) ':'
       { check_label_name lexbuf name;
         OPTLABEL name }
   | "?" (lowercase_latin1 identchar_latin1 * as name) ':'
       { warn_latin1 lexbuf;
         OPTLABEL name }
+  | raw_ident_escape (lowercase identchar * as name)
+      { LIDENT name }
```
Comment (OlivierNicole) [inline parsing/lexer.mll:413]: I was expecting to see the equivalent of this for `uppercase` and `UIDENT`. Without it, raw identifiers can only start with a lowercase character, and are a bit less general than what the RFC mentions:

> This syntax can be used anywhere that foo can be used: `~\#foo` is a labelled argument, `` `\#Foo`` is a polymorphic variant, `'\#foo` is a type variable, `\#Foo` is a module or constructor name, etc.

Arguably this is sufficient as OCaml has no capitalized keywords. I just want to make sure that’s what we want.

File: parsing/pprintast.ml [hunk 3]
```diff
@@ -134,11 +135,15 @@ type construct =
   | `nil
   | `normal
   | `simple of Longident.t
-  | `tuple ]
+  | `tuple
+  | `btrue
+  | `bfalse ]
```
Comment (OlivierNicole) [inline parsing/pprintast.ml:140]: Why is it necessary to add cases for `true` and `false` in this patch?

Comment (OlivierNicole) [reply parsing/pprintast.ml:140]: OK, I think I understand. `true` and `false` are the only keywords that can be used as constructors, so we must be wary of not printing them as escaped (since, on the other hand, raw identifiers cannot be used as constructor names).

**Review by OlivierNicole — COMMENTED** [review]
I believe the code changes of this PR are sound.

My only interrogation is at the level of the specification. As of the current state of this PR:

- Module names and value constructors can never be raw identifiers (unlike what the original RFC suggests), because raw identifiers cannot be capitalized (e.g. `\#Foo`).
- However, polymorphic variant constructors can be raw identifiers (e.g. `` `\#true``).

~~(I thought polymorphic variant constructors were necessarily capitalized, but I checked and apparently they don’t need to be, i.e. `` `x `` is valid.)~~ This is discouraged, see my message below.

As a consequence, the simple rule “raw identifiers can be used whenever any non-capitalized identifier can be used” applies. I think this simple rule should be stated somewhere (probably in the manual) so that it is clear how to use raw identifiers.

**Comment by OlivierNicole** [comment]
Using non-capitalized tag names is in fact not considered standard and may be explicitly disallowed in the future (see https://github.com/ocaml/ocaml/issues/12047). So maybe we don’t want to encourage the use of raw identifiers as tag names.

**Comment by lpw25** [comment]
I think that raw identifiers as tag names should have the same status as lowercase tag names i.e. undocumented but supported. Then if we ever remove lowercase tag names or make them official we can do the same with raw identifier tag names. For the same reasons that no one has managed to remove lowercase tag names -- they are actually widely used -- it's important to be able to use raw identifiers for them too. Otherwise when adding a keyword you risk not being able to use a library that uses that keyword as a tag name.

**Comment by OlivierNicole** [comment]
I see, that makes sense.

File: Changes [hunk 4]
```diff
@@ -93,7 +96,7 @@ Working version
   (Stephen Dolan, review by Xavier Leroy)
 
 - #11418, #11708: RISC-V multicore support.
-  (Nicolás Ojeda Bär, review by KC Sivaramakrishnan)
+  (Nicolás Ojeda Bär, review by KuC Sivaramakrishnan)
```
Comment (gasche) [inline Changes:99]: This is an unrelated typo.

File: parsing/pprintast.ml [hunk 5]
```diff
@@ -95,16 +95,17 @@ let needs_parens txt =
 let needs_spaces txt =
   first_is '*' txt || last_is '*' txt
 
-let string_loc ppf x = fprintf ppf "%s" x.txt
-
 (* add parentheses to binders when they are in fact infix or prefix operators *)
```
Comment (gasche) [inline parsing/pprintast.ml:98]: This docstring should be updated, given that the feature scope was extended to also "protect" identifiers that are also keywords. Suggestion:

```
(* Turn an arbitrary identifier name into a lexically valid identifier. *)
```

I realize as I write this that I don't know what is the right terminology for these distinct objects (the "string payload of an identifier", which I call "identifier name" here, and the "identifier as it appears in the source file", which I call a "lexically valid identifier'.)

File: parsing/pprintast.ml [hunk 6]
```diff
@@ -268,16 +273,21 @@ let iter_loc f ctxt {txt; loc = _} = f ctxt txt
 
 let constant_string f s = pp f "%S" s
 
-let tyvar ppf s =
+let tyvar_name s =
```
Comment (gasche) [inline parsing/pprintast.ml:276]: I find the new function name `tyvar_name` confusing. If I was asked to guess what is the "name" of a tyvar, I would say that `foo` is the name of the tyvar `'foo`. This function is in fact doing the opposite thing.

File: typing/typetexp.ml [hunk 7]
```diff
@@ -684,7 +686,8 @@ open Printtyp
 
 let report_error env ppf = function
   | Unbound_type_variable name ->
-      let add_name name _ l = if name = "_" then l else ("'" ^ name) :: l in
+      let add_name name _ l =
+        if name = "_" then l else Pprintast.tyvar_name name :: l in
```
Comment (gasche) [inline typing/typetexp.ml:690]: I wonder if this logic (being a no-op on `"_"`) should be added to `tyvar_name`.

**Review by gasche — COMMENTED** [review]
I looked at this PR.

I find myself confused by some of the naming choices in the code, and I think that this is because there is a lack of clarity in the terminology. When we write the OCaml expression `\#foo`, we may want to talk about `\#foo` (the identifier as it appears in source code) and `foo` (the internal payload / identifying information of the identifier). These two different things should have different names. Currently we both call them "identifiers", and then we have to remember to "protect" an identifier before printing it. This is the same distinction between `(+)` and `+`, `(let!)` and `let!`.

Then there is a similar question at the level of type variables, between `'a` and `a`. (Is it exactly the same distinction, or a distinct distinction?)

I would propose to call `\#foo` and `(+)` identifiers, and `foo` and `+` the *name* of those identifiers. This feels natural to me, and it is consistent with [the Rust documentation of raw identifiers](https://doc.rust-lang.org/rust-by-example/compatibility/raw_identifiers.html) (emphasis mine):

> This is particularly useful when Rust introduces new keywords, and a library using an older edition of Rust has a variable or function with *the same name* as a keyword introduced in a newer edition.

If we like this terminology (other proposals are welcome), we should use it in the code, in particular `protect_ident` could be `ident_of_name` or just `ident` (and be properly documented) and `tyvar_name` could be `tyvar_of_name` or just `tyvar`.

**Comment by gasche** [comment]
@OlivierNicole I remember that we discussed this PR and formed a cunning plan, but I don't remember what the plan was. You were supposed to resubmit a rebased/updated version of this PR, and I would review it; or maybe the other way around? Do you remember, which one do you prefer now?

(If you resubmit and I review, we don't need to find yet another maintainer for approval before merging. But then if you prefer to keep reviewing that is also fine by me.)

**Comment by OlivierNicole** [comment]
> @OlivierNicole I remember that we discussed this PR and formed a cunning plan, but I don't remember what the plan was. You were supposed to resubmit a rebased/updated version of this PR, and I would review it; or maybe the other way around? Do you remember, which one do you prefer now?

I can rebase and update it within a few days.

**Comment by gasche** [comment]
This has been rebased as an independent PR #12323, which is now merged. Closing. Thanks @stedolan for the nice RFC and implementation, and @OlivierNicole for going the last mile.
